*.testListenersNonSmartRoutingTerminateRandomNode

```
java.lang.AssertionError: Received event count is 0 but it is expected to be at least 10
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.client.listeners.AbstractListenersOnReconnectTest$6.run(AbstractListenersOnReconnectTest.java:418)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:967)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:984)
	at com.hazelcast.client.listeners.AbstractListenersOnReconnectTest.validateListenerFunctionality(AbstractListenersOnReconnectTest.java:414)
	at com.hazelcast.client.listeners.AbstractListenersOnReconnectTest.validateRegistrationsAndListenerFunctionality(AbstractListenersOnReconnectTest.java:377)
	at com.hazelcast.client.listeners.AbstractListenersOnReconnectTest.testListenersTerminateRandomNode(AbstractListenersOnReconnectTest.java:106)
	at com.hazelcast.client.listeners.AbstractListenersOnReconnectTest.testListenersNonSmartRoutingTerminateRandomNode(AbstractListenersOnReconnectTest.java:70)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
	at java.lang.reflect.Method.invoke(Method.java:597)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:89)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:81)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
	at java.util.concurrent.FutureTask.run(FutureTask.java:138)
	at java.lang.Thread.run(Thread.java:662)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-freestyle/13/testReport/junit/com.hazelcast.client.listeners/TopicOnReconnectTest/testListenersNonSmartRoutingTerminateRandomNode/

https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-OracleJDK8/com.hazelcast$hazelcast-client/1084/testReport/junit/com.hazelcast.client.listeners/SetItemListenerOnReconnectTest/testListenersNonSmartRoutingTerminateRandomNode/

```
java.lang.AssertionError: Received event count is 0 but it is expected to be at least 10
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.client.listeners.AbstractListenersOnReconnectTest$9.run(AbstractListenersOnReconnectTest.java:513)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:967)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:984)
	at com.hazelcast.client.listeners.AbstractListenersOnReconnectTest.validateListenerFunctionality(AbstractListenersOnReconnectTest.java:509)
	at com.hazelcast.client.listeners.AbstractListenersOnReconnectTest.validateRegistrationsAndListenerFunctionality(AbstractListenersOnReconnectTest.java:468)
	at com.hazelcast.client.listeners.AbstractListenersOnReconnectTest.testListenersTerminateRandomNode(AbstractListenersOnReconnectTest.java:108)
	at com.hazelcast.client.listeners.AbstractListenersOnReconnectTest.testListenersNonSmartRoutingTerminateRandomNode(AbstractListenersOnReconnectTest.java:72)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:89)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:81)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:745)
```

--- a/hazelcast-client/src/test/java/com/hazelcast/client/listeners/AbstractListenersOnReconnectTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/listeners/AbstractListenersOnReconnectTest.java
@@ -39,7 +39,10 @@ import org.junit.After;
 import org.junit.Test;
 
 import java.util.Collection;
+import java.util.Collections;
 import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
@@ -51,12 +54,14 @@ import static org.junit.Assert.assertTrue;
 public abstract class AbstractListenersOnReconnectTest extends ClientTestSupport {
 
     protected HazelcastInstance client;
-    protected AtomicInteger eventCount;
+    private AtomicInteger eventCount;
     private String registrationId;
     private int clusterSize;
 
     private static final int EVENT_COUNT = 10;
     private TestHazelcastFactory factory = new TestHazelcastFactory();
+    private Set<String> events = Collections.newSetFromMap(new ConcurrentHashMap<String, Boolean>());
+    private CountDownLatch eventsLatch = new CountDownLatch(1);
 
     @After
     public void tearDown() {
@@ -502,28 +507,31 @@ public abstract class AbstractListenersOnReconnectTest extends ClientTestSupport
     }
 
     private void validateListenerFunctionality() {
-        for (int i = 0; i < EVENT_COUNT; i++) {
-            produceEvent();
-        }
-
         assertTrueEventually(new AssertTask() {
             @Override
             public void run() throws Exception {
-                int count = eventCount.get();
-                assertTrue("Received event count is " + count + " but it is expected to be at least " + EVENT_COUNT,
-                        count >= EVENT_COUNT);
+                events.clear();
+                eventCount.set(0);
+                for (int i = 0; i < EVENT_COUNT; i++) {
+                    events.add(randomString());
+                }
+
+                for (String event : events) {
+                    produceEvent(event);
+                }
+
+                assertTrueAllTheTime(new AssertTask() {
+                    @Override
+                    public void run()
+                            throws Exception {
+                        int count = eventCount.get();
+                        assertEquals("Received event count is " + count + " but it is expected to stay at " + EVENT_COUNT, EVENT_COUNT,
+                                eventCount.get());
+                    }
+                }, 3);
             }
         });
 
-        assertTrueAllTheTime(new AssertTask() {
-            @Override
-            public void run()
-                    throws Exception {
-                int count = eventCount.get();
-                assertEquals("Received event count is " + count + " but it is expected to stay at " + EVENT_COUNT, EVENT_COUNT,
-                        eventCount.get());
-            }
-        }, 3);
     }
 
     private void terminateRandomNode() {
@@ -570,7 +578,15 @@ public abstract class AbstractListenersOnReconnectTest extends ClientTestSupport
 
     protected abstract String addListener();
 
-    protected abstract void produceEvent();
+    protected abstract void produceEvent(String event);
+
+    void onEvent(String event) {
+        events.remove(event);
+        eventCount.incrementAndGet();
+        if (events.isEmpty()) {
+            eventsLatch.countDown();
+        }
+    }
 
     protected abstract boolean removeListener(String registrationId);
 }

--- a/hazelcast-client/src/test/java/com/hazelcast/client/listeners/EntryListenerOnReconnectTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/listeners/EntryListenerOnReconnectTest.java
@@ -27,7 +27,7 @@ import org.junit.runner.RunWith;
 
 @RunWith(HazelcastParallelClassRunner.class)
 @Category({QuickTest.class, ParallelTest.class})
-public class EntryListenerOnReconnectTest extends AbstractListenersOnReconnectTest {
+public class EntryListenerOnReconnectTest extends AbstractListenersOnReconnectTest{
 
     private IMap<String, String> iMap;
 
@@ -35,17 +35,17 @@ public class EntryListenerOnReconnectTest extends AbstractListenersOnReconnectTe
     protected String addListener() {
         iMap = client.getMap(randomString());
 
-        final EntryAdapter<Object, Object> listener = new EntryAdapter<Object, Object>() {
-            public void onEntryEvent(EntryEvent<Object, Object> event) {
-                eventCount.incrementAndGet();
+        final EntryAdapter<String, String> listener = new EntryAdapter<String, String>() {
+            public void onEntryEvent(EntryEvent<String, String> event) {
+                onEvent(event.getKey());
             }
         };
         return iMap.addEntryListener(listener, true);
     }
 
     @Override
-    public void produceEvent() {
-        iMap.put(randomString(), randomString());
+    public void produceEvent(String event) {
+        iMap.put(event, randomString());
     }
 
     @Override

--- a/hazelcast-client/src/test/java/com/hazelcast/client/listeners/ListItemListenerOnReconnectTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/listeners/ListItemListenerOnReconnectTest.java
@@ -27,7 +27,7 @@ import org.junit.runner.RunWith;
 
 @RunWith(HazelcastParallelClassRunner.class)
 @Category({QuickTest.class, ParallelTest.class})
-public class ListItemListenerOnReconnectTest extends AbstractListenersOnReconnectTest {
+public class ListItemListenerOnReconnectTest extends AbstractListenersOnReconnectTest{
 
     private IList<String> iList;
 
@@ -37,8 +37,8 @@ public class ListItemListenerOnReconnectTest extends AbstractListenersOnReconnec
 
         ItemListener<String> listener = new ItemListener<String>() {
             @Override
-            public void itemAdded(ItemEvent item) {
-                eventCount.incrementAndGet();
+            public void itemAdded(ItemEvent<String> item) {
+                onEvent(item.getItem());
             }
 
             @Override
@@ -50,8 +50,8 @@ public class ListItemListenerOnReconnectTest extends AbstractListenersOnReconnec
     }
 
     @Override
-    public void produceEvent() {
-        iList.add(randomString());
+    public void produceEvent(String event) {
+        iList.add(event);
     }
 
     @Override

--- a/hazelcast-client/src/test/java/com/hazelcast/client/listeners/MultiMapEntryListenerOnReconnectTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/listeners/MultiMapEntryListenerOnReconnectTest.java
@@ -11,7 +11,7 @@ import org.junit.runner.RunWith;
 
 @RunWith(HazelcastParallelClassRunner.class)
 @Category({QuickTest.class, ParallelTest.class})
-public class MultiMapEntryListenerOnReconnectTest extends AbstractListenersOnReconnectTest {
+public class MultiMapEntryListenerOnReconnectTest extends AbstractListenersOnReconnectTest{
 
     private MultiMap<String, String> multiMap;
 
@@ -22,15 +22,15 @@ public class MultiMapEntryListenerOnReconnectTest extends AbstractListenersOnRec
         EntryAdapter<String, String> listener = new EntryAdapter<String, String>() {
             @Override
             public void onEntryEvent(EntryEvent<String, String> event) {
-                eventCount.incrementAndGet();
+                onEvent(event.getKey());
             }
         };
         return multiMap.addEntryListener(listener, true);
     }
 
     @Override
-    public void produceEvent() {
-        multiMap.put(randomString(), randomString());
+    public void produceEvent(String event) {
+        multiMap.put(event, randomString());
     }
 
     @Override

--- a/hazelcast-client/src/test/java/com/hazelcast/client/listeners/QueueItemListenerOnReconnectTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/listeners/QueueItemListenerOnReconnectTest.java
@@ -27,7 +27,7 @@ import org.junit.runner.RunWith;
 
 @RunWith(HazelcastParallelClassRunner.class)
 @Category({QuickTest.class, ParallelTest.class})
-public class QueueItemListenerOnReconnectTest extends AbstractListenersOnReconnectTest {
+public class QueueItemListenerOnReconnectTest extends AbstractListenersOnReconnectTest{
 
     private IQueue<String> iQueue;
 
@@ -37,8 +37,8 @@ public class QueueItemListenerOnReconnectTest extends AbstractListenersOnReconne
 
         ItemListener<String> listener = new ItemListener<String>() {
             @Override
-            public void itemAdded(ItemEvent item) {
-                eventCount.incrementAndGet();
+            public void itemAdded(ItemEvent<String> item) {
+                onEvent(item.getItem());
             }
 
             @Override
@@ -50,8 +50,8 @@ public class QueueItemListenerOnReconnectTest extends AbstractListenersOnReconne
     }
 
     @Override
-    public void produceEvent() {
-        iQueue.add(randomString());
+    public void produceEvent(String event) {
+        iQueue.add(event);
     }
 
     @Override

--- a/hazelcast-client/src/test/java/com/hazelcast/client/listeners/ReplicatedMapEntryListenerOnReconnectTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/listeners/ReplicatedMapEntryListenerOnReconnectTest.java
@@ -37,15 +37,15 @@ public class ReplicatedMapEntryListenerOnReconnectTest extends AbstractListeners
         final EntryAdapter<String, String> listener = new EntryAdapter<String, String>() {
             @Override
             public void onEntryEvent(EntryEvent<String, String> event) {
-                eventCount.incrementAndGet();
+                onEvent(event.getKey());
             }
         };
         return replicatedMap.addEntryListener(listener);
     }
 
     @Override
-    public void produceEvent() {
-        replicatedMap.put(randomString(), randomString());
+    public void produceEvent(String event) {
+        replicatedMap.put(event, randomString());
     }
 
     @Override

--- a/hazelcast-client/src/test/java/com/hazelcast/client/listeners/SetItemListenerOnReconnectTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/listeners/SetItemListenerOnReconnectTest.java
@@ -37,8 +37,8 @@ public class SetItemListenerOnReconnectTest extends AbstractListenersOnReconnect
 
         ItemListener<String> listener = new ItemListener<String>() {
             @Override
-            public void itemAdded(ItemEvent item) {
-                eventCount.incrementAndGet();
+            public void itemAdded(ItemEvent<String> item) {
+                onEvent(item.getItem());
             }
 
             @Override
@@ -49,8 +49,8 @@ public class SetItemListenerOnReconnectTest extends AbstractListenersOnReconnect
     }
 
     @Override
-    public void produceEvent() {
-        iSet.add(randomString());
+    public void produceEvent(String event) {
+        iSet.add(event);
     }
 
     @Override

--- a/hazelcast-client/src/test/java/com/hazelcast/client/listeners/TopicOnReconnectTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/listeners/TopicOnReconnectTest.java
@@ -21,16 +21,16 @@ public class TopicOnReconnectTest extends AbstractListenersOnReconnectTest {
 
         MessageListener<String> listener = new MessageListener<String>() {
             @Override
-            public void onMessage(Message message) {
-                eventCount.incrementAndGet();
+            public void onMessage(Message<String> message) {
+                onEvent(message.getMessageObject());
             }
         };
         return topic.addMessageListener(listener);
     }
 
     @Override
-    public void produceEvent() {
-        topic.publish(randomString());
+    public void produceEvent(String event) {
+        topic.publish(event);
     }
 
     @Override

